### PR TITLE
feat(plugin-workspace-tools): add flag to foreach to consider dependent workspaces

### LINKED ISSUE #1976: [Feature] Workspace foreach doesn't find dependencies
- [x] I'd be willing to implement this feature
- [ ] This feature can already be implemented through a plugin

**Describe the user story**

I have a project with the following (abbreviated) structure:

```
/package.json {
  name: "@ns/monorepo",
  workspaces: [ "common", "db", "server", "unrelated" ] }
/common/package.json { name: "@ns/common" }
/db/package.json {
  name: "@ns/db",
  dependencies: {
    "@ns/common": "workspace:common" }}
/server/package.json {
  name: "@ns/server"
  dependencies: {
    "@ns/common": "workspace:common",
    "@ns/db": "workspace:db" }}
/unrelated/package.json { name: "@ns/unrelated" }
```

So 1 project, with three workspaces, all marked as dependencies of each other.

I'd like to be able to run `build` (for typescript compilation) in `server` and have all pre-dependencies built first.

**Describe the solution you'd like**

A command like `yarn workspaces foreach run build`:

```sh
$ cd server/
$ yarn workspaces foreach -v --dependencies run build
-> built common
-> built db
-> built server
```

**Describe the drawbacks of your solution**

If different from `yarn workspaces foreach`, it may cause confusion.

**Describe alternatives you've considered**

(I thought `yarn workspaces foreach -t run build` was for my use-case but apparently I was mistaken)

**Additional context**

Discussed this with @merceyz in Discord.

### PR DESCRIPTION
**Pull request by CodeLenny**
The current behavior of `yarn workspaces foreach` first looks up workspaces under the current workspace, then evaluates dependencies.

This PR adds a `--recursive` flag, which finds workspaces anywhere in the project by evaluating `dependencies`/`devDependencies` first, before handing the projects to the same processing steps to filter (via includes/excludes) and dependency ordering (via `-t`).

Closes #1976 

**How did you fix it?**

Added `-R,--recursive` flag.  The start of `foreach` forks on the `-R` flag, between using `workspaces: []` (current behavior) and recursively using `dependencies`/`devDependencies` to find workspaces.

**Checklist**

- [x] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).

- [x] I have set the packages that need to be released for my changes to be effective.

- [x] I will check that all automated PR checks pass before the PR gets reviewed.

### CODE REVIEW

File: packages/plugin-workspace-tools/sources/commands/foreach.ts [hunk 1]
```diff
@@ -28,6 +28,30 @@ const getWorkspaceChildrenRecursive = (rootWorkspace: Workspace, project: Projec
...
+/**
+ * Find workspaces marked as dependencies/devDependencies of the current workspace recursively.
+ *
+ * @param rootWorkspace root workspace
+ * @param project project
+ *
+ * @returns all the workspaces marked as dependencies
+ */
+const getWorkspaceDependenciesRecursive = (rootWorkspace: Workspace, project: Project): Array<Workspace> => {
+  const workspaceList = [];
```
Comment (merceyz) [inline packages/plugin-workspace-tools/sources/commands/foreach.ts:40]: Should probably use a Set here and you also need to handle circular dependencies so it doesn't get stuck in an infinite loop

File: packages/plugin-workspace-tools/bin/@yarnpkg/plugin-workspace-tools.js [hunk 2]
```diff
@@ -2,21 +2,21 @@
...
...==e.invalid&&!0===t.escapeInvalid,i="";if(e.value)return(a||s)&&o.isOpenOrClose(e)?"\\"+e.value:e.value;if(e.value)return e.value;if(e.nodes)for(let t of e.nodes)i+=n(t);return i};return n(e)}},542:(e,t)=>{t.isInteger=e=>"number"==typeof e?Number.isInteger(e):"string"==typeof e&&""!==e.trim()&&Number.isInteger(Number(e)),t.find=(e,t)=>e.nodes.find(e=>e.type===t),t.exceedsLimit=(e,n,o=1,r)=>!1!==r&&(!(!t.isInteger(e)||!t.isInteger(n))&&(Number(n)-Number(e))/Number(o)>=r),t.escapeNode=(e,t=0,n)=>{let o=e.nodes[t];o&&(n&&o.type===n||"open"===o.type||"close"===o.type)&&!0!==o.escaped&&(o.value="\\"+o.value,o.escaped=!0)},t.encloseBrace=e=>"brace"===e.type&&(e.commas>>0+e.ranges>>0==0&&(e.invalid=!0,!0)),t.isInvalidBrace=e=>"brace"===e.type&&(!(!0!==e.invalid&&!e.dollar)||(e.commas>>0+e.ranges>>0==0||!0!==e.open||!0!==e.close)&&(e.invalid=!0,!0)),t.isOpenOrClose=e=>"open"===e.type||"close"===e.type||(!0===e.open||!0===e.close),t.reduce=e=>e.reduce((e,t)=>("text"===t.type&&e.push(t.value),"range"===t.type&&(t.type="text"),e),[]),t.flatten=(...e)=>{const t=[],n=e=>{for(let o=0;o<e.length;o++){let r=e[o];Array.isArray(r)?n(r,t):void 0!==r&&t.push(r)}return t};return n(e),t}},169:(e,t,n)=>{
```
Comment (merceyz) [inline packages/plugin-workspace-tools/bin/@yarnpkg/plugin-workspace-tools.js:5]: Revert the changes in this file please, building it like this is the same as releasing it

File: packages/plugin-workspace-tools/sources/commands/foreach.ts (hunk 1 above)
Comment (CodeLenny) [reply packages/plugin-workspace-tools/sources/commands/foreach.ts:44]: @merceyz I did a slightly hacky tracking of evaluated dependencies - is what I did clean enough?

Comment (merceyz) [reply packages/plugin-workspace-tools/sources/commands/foreach.ts:44]: You seem to have added it to `getWorkspaceChildrenRecursive` instead of `getWorkspaceDependenciesRecursive`

Comment (CodeLenny) [reply packages/plugin-workspace-tools/sources/commands/foreach.ts]: @merceyz This is what I get for coding right after waking up.  I've pushed a correction.

**Review by merceyz — APPROVED** [review]
LGTM :+1: 

---

I reverted the build changes as that is done when releasing, I also refactored `getWorkspaceDependenciesRecursive` to not mutate the arguments provided to it
